build/build.go

```diff
@@ -744,14 +743,60 @@ func (p *PackageData) InstallPath() string {
 	return p.PkgObj
 }
 
+// ParsedPackage is the results from building a package before it is compiled.
+type ParsedPackage struct {
+	ImportPath string // import path of package ("" if unknown)
+	Dir        string // directory containing package sources
+
+	// GoFiles is the parsed and augmented Go AST files for the package.
+	GoFiles []*ast.File
+	FileSet *token.FileSet
+	JSFiles []JSFile
+}
+
+// Imports calculates the import paths of the package's dependencies
+// based on all the imports in the augmented files.
+//
+// The given skip paths will not be returned in the results.
+// This will not return any `*_test` packages in the results.
+func (p *ParsedPackage) Imports(skip ...string) []string {
+	seen := make(map[string]struct{})
+	for _, s := range skip {
+		seen[s] = struct{}{}
+	}
+	imports := []string{}
+	for _, file := range p.GoFiles {
+		for _, imp := range file.Imports {
+			path := strings.Trim(imp.Path.Value, `"`)
+			if _, ok := seen[path]; !ok {
+				if !strings.HasSuffix(path, "_test") {
+					imports = append(imports, path)
+				}
+				seen[path] = struct{}{}
+			}
+		}
+	}
+	sort.Strings(imports)
+	return imports
+}
+
 // Session manages internal state GopherJS requires to perform a build.
 //
 // This is the main interface to GopherJS build system. Session lifetime is
 // roughly equivalent to a single GopherJS tool invocation.
 type Session struct {
-	options    *Options
-	xctx       XContext
-	buildCache cache.BuildCache
+	options *Options
+	xctx    XContext
+
+	// importPaths is a map of the resolved import paths given the srcDir
+	// and path. This is used to avoid redetermining build packages during
+	// compilation when we're looking up parsed packages.
+	importPaths map[string]map[string]string
```

Comment:
nit: If I understood this correctly:

 - The first key is the package source directory.
 - The second key is an unresolved import path from AST
 - The value is the corresponding resolved absolute import path.

If so, could you state this in the comment more explicitly?

Reply:
I tried to make it clearer. Let me know if that's still not explicit enough.

Reply:
LGTM, thanks!